cpp/include/raft/core/serialization.hpp

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/handle.hpp>
+#include <raft/detail/serialization.hpp>
+
+#include <vector>
+
+namespace raft {
+
+/**
+ * @brief Write a serializable state of an object to memory.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param[out] out a host pointer to the location where to store the state;
+ *   when `nullptr`, the actual data is not written (only the size-to-written is calculated).
+ * @param obj the object to be serialized
+ * @param args context required for serialization
+ * @return the number of bytes (to be) written by the pointer
+ */
+template <typename T, typename... ContextArgs>
+auto serialize(uint8_t* out, const T& obj, ContextArgs&&... args) -> size_t
+{
+  return detail::call_serialize<T, ContextArgs...>(out, obj, std::forward<ContextArgs>(args)...);
+}
+
+/**
+ * @brief Write a serializable state of an object to a host vector.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param obj the object to be serialized
+ * @param args context required for serialization
+ * @return the serialized state
+ */
+template <typename T, typename... ContextArgs>
+auto serialize(const T& obj, ContextArgs&&... args) -> std::vector<uint8_t>
+{
+  return detail::call_serialize<T, ContextArgs...>(obj, std::forward<ContextArgs>(args)...);
+}
+
+/**
+ * @brief Read a serializable state of an object from memory.
+ *
+ * @tparam T type of the serializable object
+ * @tparam ContextArgs types of context required for serialization
+ *
+ * @param[out] p an unitialized host pointer to a location where the object should be created.
+ * @param[in] in a host pointer to the location where the state should be read from.
+ * @param args context required for serialization
+ * @return the number of bytes read by the pointer
+ */
+template <typename T, typename... ContextArgs>
+auto deserialize(T* p, const void* in, ContextArgs&&... args) -> size_t
```

Comment:
1. What happens if deserialization fails (e.g., because the representation is not correct)?  How would you report the error?
2. It sounds like the intent is for `deserialize` to do placement new in `p`.  Is that correct?
3. Would you instead consider an interface that writes to an existing T via `T& out`?  The issue with a raw pointer interface is that users may forget the invariant "an *uninitialized* host pointer" and pass in a pointer to an existing object.  In that case, your interface can totally break any invariants of `T`, e.g., possibly leaking any heap allocations that `T` might have inside.

Reply:
1. At the moment, there are no safety checks at all. That being said, I think, we could add them in `to_bytes`/`from_bytes`, i.e. make it the responsibility of a person who adds serialization to a particular class - just throw one of the raft errors. What do you think? CC @cjnolet
2. Yes
3. I'd be fine with that, but I'm not sure how to do that better, while keeping the `serial::from_bytes` the same. Shall I add an overload that accepts `T& out`, manually calls its destructor and then passes the `&out` pointer further?.. sounds hackish.

In general, the reason why I'm hesitant to change `serial::from_bytes` to take a reference to an initialized object is that some of the classes in raft/cuml don't seem to have constructors that _do nothing_. On the one hand, the pre-initialization would likely mean some extra overhead and a minor nuisance of passing dummy arguments if there is no nullary constructor available. On the other hand, that would make impossible the value-returning overload of deserialize (`deserialize(const std::vector<uint8_t>& in, ContextArgs&&... args) -> T`) for the types that don't have the nullary constructor.

Reply:
> Shall I add an overload that accepts `T& out`, manually calls its destructor and then passes the `&out` pointer further?.. sounds hackish.

Please do not : - )  

> On the other hand, that would make impossible the value-returning overload ... for the types that don't have the nullary constructor.

It's possible if you have assignment, but ugly -- https://godbolt.org/z/774Ebfvh5 -- so I understand if you don't want to do it.

Reply:
I meant to say, it would be impossible if I changed `serial::from_bytes` to accept the reference instead of the pointer (`deserialize_raw` in your example). Also, `deserialize_using_raw` in your example does not call the destructor, does it? :) (nb my `union` trick to address this and copy elision)

Reply:
@achirkin Right, I definitely forgot the important detail of manually invoking the destructor on a placement new'd thing.